src/django_components/component_registry.py

```diff
@@ -14,49 +16,274 @@ class NotRegistered(Exception):
+    @property
+    def library(self) -> Library:
+        """
+        The template tag library with which the component registry is associated.
+        """
+        # Lazily use the default library if none was passed
+        if self._library is not None:
+            lib = self._library
+        else:
+            from django_components.templatetags.component_tags import register as tag_library
+
+            lib = self._library = tag_library
+        return lib
+
+    def register(self, name: str, component: Type["Component"]) -> None:
+        """
+        Register a component with this registry under the given name.
+
+        A component MUST be registered before it can be used in a template such as:
+        ```django
+        {% component "my_comp" %}{% endcomponent %}
+        ```
+
+        Raises `AlreadyRegistered` if a different component was already registered
+        under the same name.
+
+        Example:
+
+        ```py
+        registry.register("button", ButtonComponent)
+        ```
+        """
         existing_component = self._registry.get(name)
-        if existing_component and existing_component._class_hash != component._class_hash:
+        if existing_component and existing_component.cls._class_hash != component._class_hash:
             raise AlreadyRegistered('The component "%s" has already been registered' % name)
-        self._registry[name] = component
+
+        used_tags = ["component"]
+
+        # Keep track of which components use which tags, because multiple components may
+        # use the same tag.
+        for tag in used_tags:
+            if tag not in self._tags:
+                self._tags[tag] = set()
+            self._tags[tag].add(name)
+
+        self._registry[name] = ComponentRegistryEntry(cls=component, tags=used_tags)
 
     def unregister(self, name: str) -> None:
+        """
+        Unlinks a previously-registered component from the registry under the given name.
+
+        Once a component is unregistered, it CANNOT be used in a template anymore.
+        Following would raise an error:
+        ```django
+        {% component "my_comp" %}{% endcomponent %}
+        ```
+
+        Raises `NotRegistered` if the given name is not registered.
+
+        Example:
+
+        ```py
+        # First register component
+        registry.register("button", ButtonComponent)
+        # Then unregister
+        registry.unregister("button")
+        ```
+        """
+        # Validate
         self.get(name)
 
+        entry = self._registry[name]
+
+        # Unregister the tag from library if this was the last component using this tag
+        for tag in entry.tags:
+            # Unlink component from tag
+            self._tags[tag].remove(name)
+
+            # Cleanup
+            is_tag_empty = not len(self._tags[tag])
+            if is_tag_empty:
+                del self._tags[tag]
+
+            is_protected = tag in {"component", "#component"}
```

Comment:
Hmm... So we are bringing back #component as an inline tag available by default? I'm open to discussing this, but it doesn't match Django's defaults having tags prefixed with "#". Thoughts?

Reply:
is_protected should probably be global, and checked for in register as well, don't you think? Otherwise you could mess up the component tag as well...

Reply:
As for this

> Hmm... So we are bringing back #component as an inline tag available by default? I'm open to discussing this, but it doesn't match Django's defaults having tags prefixed with "#". Thoughts?

Let's discuss that on the next MR, which will include the tag formatter. In my first implementation of the tag formatter, I assumed that there will ALWAYS be both an "inline" and "block" forms. But since you're bringing this up, I can refactor the tag formatter such that it can define only inline or only block forms. That way, the default implementation could be Django-compatible, and users would opt-in into `#component` format.